# HG changeset patch
# User Rik <rik@octave.org>
# Date 1477607471 25200
#      Thu Oct 27 15:31:11 2016 -0700
# Node ID 62702ba67a3889abf4a1965335dd5704b5b056de
# Parent  e70551dacef6a4b0fe2d955054c98e6a1a0aabe6
__finish__.m: Change from script to function.
Changing to function avoids corner case of using 'echo on' and having
__finish__ script printed.

* __finish__.m: Change to function.  Use evalin() to call "finish"
in base workspace.

diff --git a/scripts/startup/__finish__.m b/scripts/startup/__finish__.m
--- a/scripts/startup/__finish__.m
+++ b/scripts/startup/__finish__.m
@@ -22,19 +22,22 @@
 ## path or current working directory and execute it.
 ##
 ## This function is intended to be executed upon a clean exit from Octave.
 ## This is accomplished in the system script @file{startup/octaverc} by use of
 ## the built-in function @code{atexit}.
 ## @seealso{atexit}
 ## @end deftypefn
 
-## No function declaration, this is an Octave script.  This means we are
-## still in the base workspace with access to all user variables.
+function __finish__ ()
 
-if (exist ("finish", "file"))
-  finish;  # No arg list here since finish might be a script.
-endif
+  if (exist ("finish", "file"))
+    ## Must use evalin for access to base workspace and user variables.
+    ## No argument list for finish because it might be a script, not function.
+    evalin ("base", "finish;");
+  endif
+
+endfunction
 
 
 ## No test needed for internal helper m-file.
 %!assert (1)
 
